Sources/Hedera/MirrorNodeGateway.swift

```diff
@@ -0,0 +1,114 @@
+/*
+ * ‌
+ * Hedera Swift SDK
+ *
+ * Copyright (C) 2022 - 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+import AsyncHTTPClient
+import Foundation
+import GRPC
+import HederaProtobufs
+import NIO
+
+internal struct MirrorNodeGateway {
+    internal var mirrorNodeUrl: String
+
+    private init(mirrorNodeUrl: String) {
+        self.mirrorNodeUrl = mirrorNodeUrl
+    }
+
+    internal static func forClient(_ client: Client) throws -> MirrorNodeGateway {
+        let mirrorNodeUrl = try MirrorNodeRouter.getMirrorNodeUrl(client.mirrorNetwork, client.ledgerId)
+
+        return .init(mirrorNodeUrl: mirrorNodeUrl)
+    }
+
+    internal static func forNetwork(_ mirrorNetwork: [String], _ ledgerId: LedgerId?) throws -> MirrorNodeGateway {
+        let mirrorNodeUrl = try MirrorNodeRouter.getMirrorNodeUrl(mirrorNetwork, ledgerId)
+
+        return .init(mirrorNodeUrl: mirrorNodeUrl)
+    }
+
+    internal func getAccountInfo(_ idOrAliasOrEvmAddress: String) async throws -> [String: Any] {
+        let fullApiUrl = MirrorNodeRouter.buildApiUrl(
+            self.mirrorNodeUrl, MirrorNodeRouter.accountsRoute, idOrAliasOrEvmAddress)
+
+        let responseBody = try await queryFromMirrorNode(fullApiUrl)
+
+        let jsonObject = try await deserializeJson(responseBody)
+
+        return jsonObject
+    }
+
+    internal func getContractInfo(_ idOrAliasOrEvmAddress: String) async throws -> [String: Any] {
+        let fullApiUrl = MirrorNodeRouter.buildApiUrl(
+            self.mirrorNodeUrl, MirrorNodeRouter.contractsRoute, idOrAliasOrEvmAddress)
+
+        let responseBody = try await queryFromMirrorNode(fullApiUrl)
+
+        let jsonObject = try await deserializeJson(responseBody)
+
+        return jsonObject
+    }
+
+    internal func getAccountTokens(_ idOrAliasOrEvmAddress: String) async throws -> [String: Any] {
+        let fullApiUrl = MirrorNodeRouter.buildApiUrl(
+            self.mirrorNodeUrl, MirrorNodeRouter.tokensAccountRoute, idOrAliasOrEvmAddress)
+
+        let responseBody = try await queryFromMirrorNode(fullApiUrl)
+
+        let jsonObject = try await deserializeJson(responseBody)
+
+        return jsonObject
+    }
+
+    private func queryFromMirrorNode(_ apiUrl: String) async throws -> String {
+        let httpClient = HTTPClient(eventLoopGroupProvider: .singleton)
+        defer {
+            try? httpClient.syncShutdown()
+        }
```

Comment:
I think it's generally best to keep these around if possible?

Reply:
Has to be around

```
AsyncHTTPClient/HTTPClient.swift:166: Fatal error: Client not shut down before the deinit. Please call client.shutdown() when no longer needed. Otherwise memory will leak.
```

Reply:
I meant like cache them, not leak them, like, re-use them